### PR TITLE
Clean up orphaned objects in state provider

### PR DESCRIPTION
**Pull request by LeiQL**
Fix issue #312

### CODE REVIEW

File: coa/pkg/apis/v1alpha2/providers/states/memorystate/memorystate.go [hunk 1]
```diff
@@ -231,9 +231,8 @@ func (s *MemoryStateProvider) Delete(ctx context.Context, request states.DeleteR
 	sLog.Debugf("  P (Memory State): delete state %s in namespace %s, traceId: %s", request.ID, namespace, span.SpanContext().TraceID().String())
 
 	if _, ok := s.Data[namespace]; !ok {
-		err = v1alpha2.NewCOAError(nil, fmt.Sprintf("entry '%s' is not found in namespace %s", request.ID, namespace), v1alpha2.NotFound)
-		sLog.Errorf("  P (Memory State): failed to delete %s: %+v, traceId: %s", request.ID, err, span.SpanContext().TraceID().String())
-		return err
+		sLog.Debugf("  P (Memory State): entry '%s' is not found in namespace %s, traceId: %s", request.ID, namespace, span.SpanContext().TraceID().String())
+		return nil
```
Comment (RemindD) [inline coa/pkg/apis/v1alpha2/providers/states/memorystate/memorystate.go:235]: I am OK to return 204 when deleting an object that doesn't exist but we need to have different response between successful deletion and deleting non-existent objects. Looks like the response for the above two cases are just the same.

Comment (LeiQL) [reply coa/pkg/apis/v1alpha2/providers/states/memorystate/memorystate.go:235]: This is good to know in Vendor but provider only returns errors or nil in Delete action. If we would like to handle different http status, we can only do Get + Delete in Vendors. If Get returns null, return `NoContent` (needs to be added in http status const)

File: api/pkg/apis/v1alpha1/managers/solution/solution-manager.go [hunk 2]
```diff
@@ -451,32 +474,34 @@ func (s *SolutionManager) Reconcile(ctx context.Context, deployment model.Deploy
...
-	// 	s.StateProvider.Delete(ctx, states.DeleteRequest{
-	// 		ID: deployment.Instance.ObjectMeta.Name,
-	// 		Metadata: map[string]interface{}{
-	// 			"namespace": namespace,
-	// 		},
-	// 	})
-	// } else {
-	s.StateProvider.Upsert(ctx, states.UpsertRequest{
-		Value: states.StateEntry{
+	if len(mergedState.TargetComponent) == 0 && remove {
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/managers/solution/solution-manager.go:477]: When to clean up the orphan summary?

File: coa/pkg/apis/v1alpha2/providers/states/memorystate/memorystate.go (hunk 1 above)
Comment (msftcoderdjw) [reply coa/pkg/apis/v1alpha2/providers/states/memorystate/memorystate.go:238]: The general behavior is returning 204 if deleting on non-existent objects and 200 if deletion is successful.

File: api/pkg/apis/v1alpha1/managers/jobs/jobs-manager.go [hunk 3]
```diff
@@ -282,16 +282,28 @@ func (s *JobsManager) HandleHeartBeatEvent(ctx context.Context, event v1alpha2.E
 		namespace = "default"
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/managers/jobs/jobs-manager.go:282]: I think we need add some E2E test cases to check the logic of cleaning up orphaned objects, e.g. re-create object with same name.

Comment (LeiQL) [reply api/pkg/apis/v1alpha1/managers/jobs/jobs-manager.go:282]: Added. Wait for test results.

File: k8s/reconcilers/deployment.go [hunk 4]
```diff
@@ -345,6 +345,13 @@ func (r *DeploymentReconciler) AttemptRemove(ctx context.Context, object Reconci
 		if object.GetStatus().ProvisioningStatus.Status == string(utilsmodel.ProvisioningStatusFailed) {
 			r.delayFunc(r.deleteSyncDelay)
 		}
+
+		// TODO: handle crash consistency that finalizer removal fails
+		err = r.deleteDeploymentSummary(ctx, object)
+		if err != nil {
+			return metrics.StatusUpdateFailed, ctrl.Result{}, err
```
Comment (msftcoderdjw) [inline k8s/reconcilers/deployment.go:352]: Define new metrics error.

File: test/integration/scenarios/03.basicWithNsDelete/verify/manifest_test.go [hunk 5]
```diff
@@ -304,6 +304,51 @@ func TestBasic_InstanceDeletion(t *testing.T) {
...
+			Group:    "solution.symphony",
+			Version:  "v1",
+			Resource: "instances",
+		}).Namespace("default").List(context.Background(), metav1.ListOptions{})
+		require.NoError(t, err)
+
+		require.Len(t, resources.Items, 1, "there should be only one instance")
+
+		status := getStatus(resources.Items[0])
+		fmt.Printf("Current instance status: %s\n", status)
```
Comment (msftcoderdjw) [inline test/integration/scenarios/03.basicWithNsDelete/verify/manifest_test.go:341]: We need to verify the status details. Cleaning up summary is to make sure the previous status won't be leaked after recreation.

**Review by msftcoderdjw — CHANGES_REQUESTED** [review]

**Review by msftcoderdjw — APPROVED** [review]

**Pull request merged**